[FEATURE] Implémentation de la nouvelle barre de navigation (PIX-1071).

alexandrecoin opened the pull request:
## :unicorn: Problème

Le site vitrine change de peau, donc la barre de navigation en haut doit faire peau neuve.

Attendu : 

https://app.abstract.com/projects/c43b8749-ba3a-474e-a6c5-dbf76bade2ee/branches/0f98f94c-8f44-4d4f-a988-72be104570aa/commits/latest/files/7CDCEB46-2B82-4218-B719-A559D817B735/layers/22F59764-AE99-4E92-8376-001C80AC08CF?collectionId=92047f7c-2ee2-4adc-9227-9b63f56bf55a&collectionLayerId=d4e6d1eb-18e3-452c-baf3-158576a9acfd

## :robot: Solution

- Remplacement des items de navigation de l'ancienne nav par les nouveaux provenant de Prismic
- Ajout d'une dropdown de sélection des éléments de sous-navigation.

## :rainbow: Remarques

- complication pour détecter si le bouton de la dropdown doit avoir le style actif: vu avec Quentin pour ne mettre le style actif que sur la sous-navigation et pas le bouton dropdown parent
- ajout d'un attribut `dropdown-index` pour défférencier les dropdowns et avoir les dropdown indépendants les uns des autres
- Vu la nav avec Quentin, globalement OK mais shadow de la nav à revoir (pas bloquant, ça peut faire l'objet d'un prochain ticket si celui-ci est mergé)

## :sparkles: Review App
https://pix-site-review-pr166.osc-fr1.scalingo.io

alexandrecoin committed:
Create navigation logo zone

jonasgrilleres committed:
Filter old navigation

Remove old navigation header display
Keep navigation items because used in page footer

alexandrecoin committed:
Display logos zone in header

alexandrecoin committed:
Update burger menu for small screens

Commit:
Remove redundant grey color

Commit:
Add navigation-zone to mainNavigation

alexandrecoin committed:
Enable image click in logo zone

Commit:
Display actions-zone in navigation

Commit:
Add actions zone to burger menu

Commit:
Display navigation dropdown

alexandrecoin committed:
Style dropdown and navigation zone

alexandrecoin committed:
Get subnavigation links

alexandrecoin committed:
Style navigation zone

Commit:
Handle multiple dropdowns in navigation

celineung commented on components/NavigationSliceZone.vue:
```diff
@@ -38,7 +38,15 @@ export default {
+    ...mapState(['mainNavigation']),
+    burgerMenuLinks() {
+      return [
+        ...this.mainNavigation.data.body[1].items,
```
**suggestion** Voir ici pour se détacher de l'index du tableau

MelanieMEB commented on components/slices/NavigationZone.vue:
```diff
@@ -0,0 +1,145 @@
+  },
+  computed: {
+    navigationLinks() {
+      const linkByGroup = {}
+      const links = []
+      this.slice.items.forEach((item) => {
+        if (item.group.length > 0) {
+          const groupName = item.group[0].text
+          if (linkByGroup[groupName] == null) {
+            linkByGroup[groupName] = []
+          }
+          linkByGroup[groupName].push(item)
+        }
+      })
+
+      const groupAlreadyInNavigation = []
+      this.slice.items.forEach((item) => {
+        if (item.group.length > 0) {
+          const groupName = item.group[0].text
+          if (!groupAlreadyInNavigation.includes(groupName)) {
+            links.push({
+              name: groupName,
+              subNavigationLinks: linkByGroup[groupName],
+            })
+            groupAlreadyInNavigation.push(groupName)
+          }
+        } else {
+          links.push(item)
+        }
+      })
+      return links
+    },
```
```suggestion
      const links = []
      this.slice.items.forEach((item) => {
        if (item.group.length > 0) {
          const groupName = item.group[0].text
          const alreadyGroupsInLinks = links.find((link)=> link.name==groupName && link.subNavigationLinks)
          if(!alreadyGroupsInLinks) {
            links.push({
              name: groupName,
              subNavigationLinks: [item],
            })
          } else {
            alreadyGroupsInLinks.subNavigationLinks.push(item)
          }
        } else {
          links.push(item);
        }
      })
      return links
    },
```

MelanieMEB reviewed (commented):
Proposition d'amélioration. Sinon ça me parait bien. A tester sous IE avec Phili

MelanieMEB commented on components/slices/NavigationZone.vue:
```diff
@@ -0,0 +1,137 @@
+        border-top: none;
+        border-bottom: none;
+        background-color: transparent;
+        outline: none;
```
Je propose de retirer cette ligne pour avoir le focus visible sur le group en navigation clavier

MelanieMEB reviewed (changes requested):
Problème de sous-menu sous IE : Impossible de cliquer sur Enseignement Sup. Je pense que le clic pour sortir est pris au dessus du clic pour le lien.

celineung commented on components/BurgerMenuNav.vue:
```diff
@@ -96,95 +61,14 @@ export default {
 
-  .nav-top a:hover,
-  .nav-bottom a:hover {
+  .nav-top a:hover {
```
```
.nav-top {
  &> a:hover {}
}
```

celineung commented on components/BurgerMenuNav.vue:
```diff
@@ -200,16 +84,14 @@ export default {
 
-.bm-burger-button {
-  @include device-is('desktop') {
+.navigation-slice-zone .bm-burger-button {
```
```
.navigation-slice-zone {
  &> bm-burger-button {}
}
```

MelanieMEB committed:
Inverse link and li to have a large button for dropdown

Commit:
Fix dropdown navigation click on IE

Problem: on IE, it was not possible to navigate to a link in the navigation dropdown
Solution: extract the dropdown menu from the button

sbedeau commented on components/BurgerMenuNav.vue:
```diff
@@ -61,9 +27,8 @@ export default {
-  padding-top: 0;
 
-  @include device-is('desktop') {
+  @include device-is('large-screen') {
```
Pourquoi large-screen plutôt que desktop ?

celineung replied:
d'après les maquettes, en desktop on a le burger menu
https://app.abstract.com/projects/c43b8749-ba3a-474e-a6c5-dbf76bade2ee/branches/0f98f94c-8f44-4d4f-a988-72be104570aa/collections/92047f7c-2ee2-4adc-9227-9b63f56bf55a

sbedeau replied:
Ça me semble être une mauvaise pratique d'avoir un burger menu en mode desktop

sbedeau commented on components/NavigationDropdown.vue:
```diff
@@ -0,0 +1,112 @@
+        if (
+          event.target.getAttribute('dropdown-index') !== this.dropdownIndex
+        ) {
+          console.log(event.target)
```
Est-ce qu'on a vraiment besoin de ce console.log ? 
Pourquoi le linter n'a pas réagit ?

celineung replied:
c'est sur un commit de WIP, je debuggais le problème d'affichage sur IE

sbedeau commented on components/NavigationDropdown.vue:
```diff
@@ -0,0 +1,112 @@
+    0px 8px 32px 0px rgba(0, 0, 0, 0.06);
+  color: $grey-9;
+  font-size: 0.875rem;
+  font-weight: 400;
```
```suggestion
  font-weight: $font-normal;
```

sbedeau commented on components/NavigationDropdown.vue:
```diff
@@ -0,0 +1,112 @@
+
+<style scoped lang="scss">
+.navigation-dropdown {
+  box-shadow: 0px 24px 32px 0px rgba(0, 0, 0, 0.03),
+    0px 8px 32px 0px rgba(0, 0, 0, 0.06);
```
```suggestion
  box-shadow: 0 24px 32px 0 rgba(0, 0, 0, 0.03),
    0 8px 32px 0 rgba(0, 0, 0, 0.06);
```

sbedeau commented on components/NavigationSliceZone.vue:
```diff
@@ -0,0 +1,96 @@
+<template>
+  <div class="navigation-slice-zone__wrapper">
+    <div class="navigation-slice-zone">
```
Selon BEM, il faudrait plutôt 
```suggestion
  <div class="navigation-slice-zone">
    <div class="navigation-slice-zone__wrapper">
```

sbedeau commented on components/NavigationSliceZone.vue:
```diff
@@ -0,0 +1,96 @@
+
+<style lang="scss">
+.navigation-slice-zone__wrapper {
+  display: grid;
+  grid-template-columns: 1.2fr repeat(12, 1fr) 1.2fr;
+  grid-template-areas: '. navigation navigation navigation navigation navigation navigation navigation navigation navigation navigation navigation navigation .';
```
Pourquoi utiliser une grid ?

celineung replied:
Pour avoir la navigation de la même taille que les éléments de la page

sbedeau replied:
Je pense que ça amène un couplage fort sur le design et un couplage complexe dans le code. Je suggère de privilégier le flex pour garantir une maintenabilité simple

celineung replied:
Mais en utilisant flex dans la navigation et grid dans les slices de page, ce n'est pas possible de les aligner

sbedeau commented on components/slices/ActionsZone.vue:
```diff
@@ -0,0 +1,67 @@
+      font-size: 0.875rem;
+      color: $grey-9;
+      font-family: $font-roboto;
+      font-weight: 500;
```
```suggestion
      font-weight: $font-medium;
```

sbedeau commented on components/NavigationDropdown.vue:
```diff
@@ -0,0 +1,112 @@
+  position: absolute;
+  left: -50px;
+  border-radius: 3px;
+  background: white;
```
```suggestion
  background: $white;
```

sbedeau commented on components/slices/LogosZone.vue:
```diff
@@ -0,0 +1,58 @@
+  },
+  methods: {
+    hasLink(item) {
+      return item.url.link_type !== 'Any'
```
Pour ma culture, c'est quoi Any ? Pourquoi Any plutôt que `!== null` ?

celineung replied:
Lorsque l'on choisit un lien sur Prismic, on choisit le type (document, lien externe, etc).
Du coup, Prismic ne renvoit jamais `null` mais `Any` s'il n'y a pas de lien.

Commit:
Style current active navigation link

sbedeau commented on components/slices/LogosZone.vue:
```diff
@@ -0,0 +1,58 @@
+
+    img {
+      height: 50px;
+      margin: 15px 0 9px 0;
```
Je pense que les margins asymétriques peuvent poser problème à la généricité du composant. 
Est-ce qu'on ne devrait pas plutôt imposer des logos sont un certain format via des margins symétriques fixes ?

sbedeau commented on components/slices/NavigationZone.vue:
```diff
@@ -0,0 +1,138 @@
+  computed: {
+    navigationLinks() {
+      const links = []
+      this.slice.items.forEach((item) => {
+        if (item.group.length > 0) {
+          const groupName = item.group[0].text
+          const groupAlreadyInLinks = links.find(
+            (link) => link.name === groupName && link.subNavigationLinks
+          )
+          if (!groupAlreadyInLinks) {
+            links.push({
+              name: groupName,
+              subNavigationLinks: [item],
+            })
+          } else {
+            groupAlreadyInLinks.subNavigationLinks.push(item)
+          }
+        } else {
+          links.push(item)
+        }
+      })
+      return links
+    },
```
La complexité cyclomatique me semble très forte ici. Est-ce qu'on pourrait faire un découpage en courtes fonctions locales explicitement nommées ?

sbedeau commented on components/BurgerMenuNav.vue:
```diff
@@ -59,159 +26,77 @@ export default {
-    display: none;
+  .bm-cross {
+    background: $grey-1;
+    height: 24px !important;
```
En général, `!important` est un bad smell. Comment est-ce qu'on peut l'enlever ?
```suggestion
    height: 24px;
```

celineung replied:
Je t'avoue ne pas avoir vu le `important` qui trainait.
En regardant de plus près, le `important` a été rajouté car le composant `vue-burger-menu` fait des manipulations directement sur le `style` de l'élément sans passer par une classe css.
Je propose de voir cela lors de la refonte du burger-menu.

sbedeau commented on components/slices/NavigationZone.vue:
```diff
@@ -0,0 +1,138 @@
+      margin: 0 4px;
+      font-family: $font-roboto;
+      font-size: 14px;
+      font-weight: 500;
```
```suggestion
      font-weight: $font-medium;
```

sbedeau reviewed (changes requested):
Retour func : il y a une ombre visible de chaque côté du menu que je n'ai pas trouvé dans la maquette

celineung commented:
> Retour func : il y a une ombre visible de chaque côté du menu que je n'ai pas trouvé dans la maquette

C'est fixé =)

Commit:
Fix organisation-nav display after rebase

sbedeau commented on components/slices/NavigationZone.vue:
```diff
@@ -0,0 +1,159 @@
+    this.links = []
+  }
+
+  addNavigationLink(navigationLink) {
```
Est-ce que c'est bien une méthode locale ?
```suggestion
  _addNavigationLink(navigationLink) {
```

celineung replied:
Non, c'est une méthode publique

sbedeau commented on components/slices/NavigationZone.vue:
```diff
@@ -0,0 +1,159 @@
+    this.links.push(navigationLink)
+  }
+
+  addSubNavigationLink(groupName, subNavigationLink) {
```
Est-ce que c'est bien une méthode locale ?
```suggestion
  _addSubNavigationLink(groupName, subNavigationLink) {
```

celineung replied:
Non, c'est une méthode publique

sbedeau commented on components/slices/NavigationZone.vue:
```diff
@@ -0,0 +1,159 @@
+          navigation.addNavigationLink(item)
+        }
+      })
+      return navigation.links
```
Pourquoi est-ce qu'on a besoin d'un objet ? Il ne contient que links d'ailleurs
```suggestion
      return links
```

celineung replied:
J'ai trouvé plus clair d'exporter la logique d'ajout dans le tableau dans un objet

sbedeau commented on components/slices/NavigationZone.vue:
```diff
@@ -0,0 +1,159 @@
+    this.links = []
+  }
+
+  addNavigationLink(navigationLink) {
+    this.links.push(navigationLink)
+  }
+
+  addSubNavigationLink(groupName, subNavigationLink) {
+    const group = this._groupAlreadyInLinks(groupName)
+    if (!group) {
+      this._createNewGroup(groupName, [subNavigationLink])
+    } else {
+      group.subNavigationLinks.push(subNavigationLink)
+    }
+  }
+
+  _groupAlreadyInLinks(groupName) {
+    return this.links.find(
+      (link) => link.name === groupName && link.subNavigationLinks
+    )
+  }
+
+  _createNewGroup(groupName, subNavigationLinks) {
+    this.links.push({
+      name: groupName,
+      subNavigationLinks,
+    })
+  }
+}
```
Je propose de mettre ces méthodes en fin de script pour suivre la step down rule et mettre en avant le script du fichier plutôt que les outils qui le compoosent

ghost commented:
Quelques commentaires ont aussi été rajoutés dans la review précédente

celineung commented:
> Problème de sous-menu sous IE : Impossible de cliquer sur Enseignement Sup. Je pense que le clic pour sortir est pris au dessus du clic pour le lien.

c'est fixé =)

MelanieMEB commented on components/BurgerMenuNav.vue:
```diff
@@ -2,55 +2,22 @@
         </li>
       </ul>
     </div>
+    g
```
Un g perdu :)
